### PR TITLE
Improve scaling speed and some cleanup

### PR DESCRIPTION
**Pull request by gaurav8297**
## Benchmarks

For unpartitioned case:

Before:
```
trino:insert_demo> insert into lineitem_test_unpart select * from lineitem_shipdate;
INSERT: 6733485888 rows

Query 20230620_050123_00012_e44xz, FINISHED, 7 nodes
Splits: 7,486 total, 7,486 done (100.00%)
7:24 [6.73B rows, 182GB] [15.2M rows/s, 421MB/s]
```

After:
```
trino:insert_demo> insert into lineitem_test_unpart select * from lineitem_shipdate;
INSERT: 6733485888 rows

Query 20230620_050758_00006_8c6sa, FINISHED, 7 nodes
Splits: 7,656 total, 7,656 done (100.00%)
4:31 [6.73B rows, 182GB] [24.8M rows/s, 689MB/s]
```

For partitioned case:

Before:
```
trino:insert_demo> insert into lineitem_part_target_shipmode select * from lineitem_part_3_skewed_partitions;
INSERT: 3085776220 rows

Query 20230620_052319_00017_e44xz, FINISHED, 7 nodes
Splits: 2,243 total, 2,243 done (100.00%)
3:43 [3.09B rows, 83.3GB] [13.8M rows/s, 382MB/s]
```

After:
```
trino:insert_demo> insert into lineitem_part_target_shipmode select * from lineitem_part_3_skewed_partitions;
INSERT: 3085776220 rows

Query 20230624_232648_00006_z65ur, FINISHED, 7 nodes
Splits: 2,242 total, 2,242 done (100.00%)
2:16 [3.09B rows, 83.3GB] [22.8M rows/s, 629MB/s]
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 1]
```diff
@@ -79,11 +79,11 @@ public class TaskManagerConfig
     private Duration interruptStuckSplitTasksDetectionInterval = new Duration(2, TimeUnit.MINUTES);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:79]: provide rationale in commit message

File: core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java [hunk 2]
```diff
@@ -76,9 +75,7 @@ private int getNextWriterIndex()
         // This also mean that we won't scale local writers if the writing speed can cope up
         // with incoming data. In another word, buffer utilization is below 50%.
         if (writerCount < buffers.size() && memoryManager.getBufferedBytes() >= maxBufferedBytes / 2) {
-            long physicalWrittenBytes = physicalWrittenBytesSupplier.get();
-            if ((physicalWrittenBytes - lastScaleUpPhysicalWrittenBytes) >= writerCount * writerMinSize) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterExchanger.java:80]: is `lastScaleUpPhysicalWrittenBytes` still needed? Can we remove?

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 3]
```diff
@@ -154,16 +155,16 @@ public FeaturesConfig setScaleWriters(boolean scaleWriters)
...
     {
-        return writerMinSize;
+        return writerScalingMinDataProcessed;
     }
 
-    @Config("writer-min-size")
-    @ConfigDescription("Target minimum size of writer output when scaling writers")
-    public FeaturesConfig setWriterMinSize(DataSize writerMinSize)
+    @Config("writer-scaling-min-data-processed")
+    @ConfigDescription("Minimum amount of uncompressed output data processed by writers before scaling can happen")
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:164]: maybe `before writer scaling can happen`

File: core/trino-main/src/main/java/io/trino/operator/DriverContext.java [hunk 4]
```diff
@@ -266,6 +266,16 @@ public CounterStat getOutputPositions()
         return new CounterStat();
     }
 
+    public long getWrittenInputDataSize()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/DriverContext.java:269]: maybe `getWrittenInputDataSize` -> `getWrittenDataSize`

File: core/trino-main/src/main/java/io/trino/operator/OperatorContext.java [hunk 5]
```diff
@@ -246,6 +247,11 @@ public void setFinishedFuture(ListenableFuture<Void> finishedFuture)
         checkState(this.finishedFuture.getAndSet(requireNonNull(finishedFuture, "finishedFuture is null")) == null, "finishedFuture already set");
     }
 
+    public void recordWriterInputDataSize(long sizeInBytes)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/OperatorContext.java:250]: Do we need extra methods for this? I think this value is just same as `inputDataSize` for `TableWriter`

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 6]
```diff
@@ -36,6 +36,7 @@
 import java.util.OptionalInt;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:36]: please provide rationale in commit message. Also, why we need new property?

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 7]
```diff
@@ -90,7 +90,7 @@ public class TaskManagerConfig
     // on the worker nodes due to more available processors, the default value could be above 1. Therefore,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:90]: provide rationale in commit msg

File: core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java [hunk 8]
```diff
@@ -347,9 +348,15 @@ private void rebalanceBasedOnTaskBucketSkewness(
                     break;
                 }
 
+                // Rebalance partition only once in a single cycle
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java:351]: Why? Add comment

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java (hunk 1 above)
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:79]: comment: some of the writers will do CPU and some IO, hence we need more writers

File: core/trino-main/src/main/java/io/trino/FeaturesConfig.java [hunk 9]
```diff
@@ -72,7 +72,7 @@ public class FeaturesConfig
 
     private boolean redistributeWrites = true;
     private boolean scaleWriters = true;
-    private DataSize writerMinSize = DataSize.of(32, DataSize.Unit.MEGABYTE);
+    private DataSize writerScalingMinDataProcessed = DataSize.of(100, DataSize.Unit.MEGABYTE);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/FeaturesConfig.java:75]: maybe, 120MB, 150MB (assuming with 100 you get 20MB physical)

File: docs/src/main/sphinx/admin/properties-writer-scaling.rst [hunk 10]
```diff
@@ -61,3 +61,15 @@ another writer is eligible to be added. Each writer task may have multiple
...
+.. warning::
+    This property is deprecated now. Please use ``writer-scaling-min-data-processed``.
+
+``writer-scaling-min-data-processed``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+* **Type:** :ref:`prop-type-data-size`
+* **Default value:** ``100MB``
+
+The minimum amount of uncompressed data that must be processed by a writer
+before another writer is eligible to be added. This can be specified on a
```
Comment (sopel39) [inline docs/src/main/sphinx/admin/properties-writer-scaling.rst:74]: nit: `is eligible to be added` -> can be added`?

File: core/trino-main/src/test/java/io/trino/operator/exchange/TestLocalExchange.java [hunk 11]
```diff
@@ -1176,4 +1155,9 @@ public static long retainedSizeOfPages(int count)
     {
         return RETAINED_PAGE_SIZE.toBytes() * count;
     }
+
+    public static long sizeOfPages(int count)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/exchange/TestLocalExchange.java:1129]: private?

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java (hunk 7 above)
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:90]: same comment as before

File: core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java [hunk 12]
```diff
@@ -347,9 +348,17 @@ private void rebalanceBasedOnTaskBucketSkewness(
                     break;
                 }
 
+                // Rebalance partition only once in a single cycle. Otherwise, rebalancing will happen quite
+                // aggressively in the early stage of write, while it is not required. Thus, it can have an impact on
+                // output file sizes and resource usage.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java:353]: nit: `output file sizes and resource usage (produced files can be small and memory usage might be higher)`.

**Review by sopel39 — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java [hunk 13]
```diff
@@ -1471,11 +1471,13 @@ default Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session,
         return Optional.empty();
     }
 
+    @Deprecated
```
Comment (sopel39) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: maybe add a comment that it's now used only in test code and is not needed for any engine logic cc @findepi

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: Unused methods should be removed from here.

For test-only functionality, this should be handled purely on the test side

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: i don't see a reason to keep unused method in ConnectorMetadata.
if someone wants to support two versions at a time, they can still compile their `ConnectorMetadata` impl (just without `@Override` annotation)

anyway, don't feel strongly about one version of not, just please do follow up. you should be able to create commit/PR without waiting for anything.

Comment (sopel39) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: @findepi I think we have one-version-backward compatibility policy

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: @gaurav8297 can you put up a PR that will remove these deprecated methods? to be merged after 422 goes out

**Comment by piotrolchawa-f3** [comment]
Hey @gaurav8297! Nice one! I was wondering about the benchmarks that you've conducted.

1. What is the Trino setup? What's the catalogue you use?
2. How does the table look like?
3. What size of servers do you use?
4. What connector do you use?

I'm asking because these tests seem great for me even before the change. We're trying to insert with benthos, and Trino chokes on 1000 rows/sec. Ofc our coordinator is kind of small (16vcpus, 32Gb ram), but I'd expect it to be able to ingest more.

**Comment by piotrolchawa-f3** [comment]
Hey @sopel39 , thanks for the quick response!

Our setup is:
1. Benthos ingests events from a message queue and `INSERT`s them in micro batches of 1000 using Trino.
2. Trino is using `blackhole` connector (to eliminate storage performance issues. But the target connector is Iceberg, on which we see exactly same performance).
3. Coodinator is running on 16vcpu, 32GB ram, there are two workers (we tried increasing number of workers, it changes nothing).
4. The query for insertion is: `INSERT INTO iceberg.test.my_table(id, date_0, tenant_id, comment_1, comment_2, is_t, code, is_s, date_1, round, date_2) VALUES (X, Y, Z....), (X1, Y1, Z1, ....)` We have a single insert query with 1000 rows.
5. Example row insertion: `INSERT INTO iceberg.test.my_table (id, date_0, tenant_id, comment_1, comment_2, is_t, code, is_s, date_1, round, date_2)
VALUES ('dm4X', from_iso8601_timestamp('2023-06-27'), CAST ('fa51f2df-6645-47fd-9209-198a045fbe9a' AS UUID), '11o', '15072', CAST (false AS BOOLEAN), 'EyGuVi9n', CAST (true AS BOOLEAN), from_iso8601_timestamp('2023-06-27'), TRY_CAST(NULL AS INT), from_iso8601_timestamp('2023-06-27'));`
6. Removing CASTs sped things a little, but not in any significant way.

So in our investigation we've discovered that PLANNING phase takes a long time:

```
  "queryStats" : {
    "createTime" : "2023-07-19T09:56:34.280Z",
    "executionStartTime" : "2023-07-19T09:56:34.388Z",
    "lastHeartbeat" : "2023-07-19T09:56:38.810Z",
    "endTime" : "2023-07-19T09:56:38.809Z",
    "elapsedTime" : "4.53s",
    "queuedTime" : "183.44us",
    "resourceWaitingTime" : "107.82ms",
    "dispatchingTime" : "49.48us",
    "executionTime" : "4.42s",
    "analysisTime" : "107.77ms",
    "planningTime" : "3.77s",
    "planningCpuTime" : "3.72s",
    "finishingTime" : "356.31us",
    "totalTasks" : 4,
    "runningTasks" : 0,
    "completedTasks" : 4,
    "failedTasks" : 0,
```
And if we zoom in on the PLANNER:

```
    "optimizerRulesSummaries" : [ {
      "rule" : "io.trino.sql.planner.iterative.rule.ExpressionRewriteRuleSet.ValuesExpressionRewrite",
      "invocations" : 104,
      "applied" : 2,
      "totalTime" : 3692627565,
      "failures" : 0
    }
```

which is 3.69s.


On the other hand these are the results of the query you've run for this PR:

```
trino:test> insert into table2 select * from table1;
INSERT: 22721776 rows

Query 20230719_093535_00048_afzcf, FINISHED, 3 nodes
Splits: 34 total, 34 done (100.00%)
18.04 [22.7M rows, 600MB] [1.26M rows/s, 33.3MB/s]
```
 
 Which looks fine, so probably it's for different reasons. Do you want us to create a separate issue for this?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java (hunk 13 above)
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:1474]: looks like this is part of https://github.com/trinodb/trino/pull/18561

**Comment by sopel39** [comment]
> Which looks fine, so probably it's for different reasons. Do you want me to create a separate issue for this?

Please create an issue for this

**Pull request merged**